internal/agent/upgrade.go

```diff
@@ -60,16 +62,21 @@ func Upgrade(
 		return err
 	}
 
-	err = upgradeSpec.Sanitize()
-	if err != nil {
-		return err
-	}
-
-	upgradeAction := action.NewUpgradeAction(c, upgradeSpec)
-
-	err = upgradeAction.Run()
-	if err != nil {
-		return err
+	if boardutils.GetAndroidBoardModel() == constants.QCS6490 {
+		if upgradeSpec.Active.Source == nil || upgradeSpec.Active.Source.IsEmpty() {
+			return fmt.Errorf("active source is empty. For QCS6490 boards, you need to specify the active source by using the ---source flag")
+		}
+		if err = action.NewUpgradeQCS6490Action(c, upgradeSpec).Run(); err != nil {
+			return err
+		}
+	} else {
+		err = upgradeSpec.Sanitize()
+		if err != nil {
+			return err
+		}
+		if err = action.NewUpgradeAction(c, upgradeSpec).Run(); err != nil {
+			return err
+		}
```

Comment:
I think it would be better to move this control flow inside NewUpgradeAction, or some wrapper around it

Reply:
yeah problem is the types no? If newupgradeAction can return either a UpgradeAction or Upgradeqcs6490Action then we need to start asbtracting the types and casting it to the final type and such

I really dislike that in go :(

Reply:
Also there is the issue with circular deps, there are some things that bring other things and other things which results in an endless circular dependency hell :sob:

Reply:
I don't think the types should be a problem if we use generics ... but these changes can be done on a different PR, I don't want to block the merging

Reply:
noh, its mainly about circular deps between the actions, the utils, and the types pkg folders, I had a bad time ttrying to stop everything depending on everything :(

Reply:
I see, yeah, if we have circular deps that's a bigger issue :/